lib/Optimizer/TrainingPreparation.cpp

```diff
@@ -0,0 +1,118 @@
+                              Tensor *tensor) {
+  switch (node->getKind()) {
+  case Kinded::Kind::ConvolutionNodeKind: {
+    ConvolutionNode *CN = llvm::dyn_cast<ConvolutionNode>(node);
+    if (CN->getFilter() == input) {
+      ShapeNHWC idim = ShapeNHWC(CN->getInput().dims());
+      ShapeHW kdim(CN->getKernels());
+      size_t fanIn = kdim.height * kdim.width * idim.c;
+      tensor->init(Tensor::InitKind::Xavier, fanIn, F->getPRNG());
+    } else if (CN->getBias() == input) {
+      tensor->init(Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+  } break;
+  case Kinded::Kind::BatchNormalizationNodeKind: {
+    BatchNormalizationNode *BN = llvm::dyn_cast<BatchNormalizationNode>(node);
+    if (BN->getScale() == input) {
+      tensor->init(glow::Tensor::InitKind::Zero, 0, F->getPRNG());
+    } else if (BN->getBias() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    } else if (BN->getMean() == input) {
+      tensor->zero();
+    } else if (BN->getVar() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 1.0, F->getPRNG());
+    }
+  } break;
+  case Kinded::Kind::FullyConnectedNodeKind: {
+    FullyConnectedNode *FCN = llvm::dyn_cast<FullyConnectedNode>(node);
+    if (FCN->getWeights() == input) {
+      auto in = FCN->getInput();
+      tensor->init(glow::Tensor::InitKind::Zero, in.dims()[1], F->getPRNG());
+    } else if (FCN->getBias() == input) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+  } break;
+  default:
+    break;
+  }
+}
+} // namespace
+
+TensorInitializer getDefaultTensorInitializer() {
+  return defaultTensorInitializer;
+}
+
+void prepareFunctionForTraining(
+    Function *F, PlaceholderBindings &bindings,
+    const std::unordered_set<std::string> &legalOutputNames,
+    TensorInitializer &&initializer) {
+
+  auto &nodes = F->getNodes();
+  // Lookup all nodes, skip Storage types, enumerate inputs,
+  // replace Constant type with trainable Placeholders except special cases,
+  // like BatchNormalization inputs (mean and variance). In special cases
+  // replace Constant type with non-trainable Placeholders.
+  for (auto &node : nodes) {
+    // Skip storages.
+    if (Storage::classof(&node)) {
+      continue;
+    }
+
+    // Skip output node(s).
+    if (legalOutputNames.count(node.getName())) {
+      continue;
+    }
+
+    for (unsigned idx = 0; idx < node.getNumInputs(); ++idx) {
+      auto INV = node.getNthInput(idx);
+      auto *IN = INV.getNode();
+      if (Constant *C = llvm::dyn_cast<Constant>(IN)) {
+        bool isTrainable = true;
+        switch (node.getKind()) {
+          // Special case. Mean and Variance inputs are not trainable.
+        case Kinded::Kind::BatchNormalizationNodeKind: {
+          if (BatchNormalizationNode *BN =
+                  llvm::dyn_cast<BatchNormalizationNode>(&node)) {
+            if (INV == BN->getMean() || INV == BN->getVar()) {
```

Comment:
You can simply check if `idx == BatchNormalizationNode::MeanIdx` etc. instead of all this logic of casting the BN and getting the actual inputs.